### PR TITLE
Fix RSA OAEP set/get label for legacy engine

### LINKED ISSUE #21335: Segmentation fault when calling EVP_PKEY_CTX_get0_rsa_oaep_label with legacy key in 3.0.9
OpenSSL 3.0.9

We are utilizing an engine provided by a third party to retrieve keys from a Hardware Secure Module (HSM). The keys are in a legacy format. However, after updating from version 3.0.8 to 3.0.9, the following code with such a key is causing a segmentation fault:
```
unsigned char *oaep_label = NULL;
auto labellen = EVP_PKEY_CTX_get0_rsa_oaep_label(ctx, &oaep_label);
```

The line causing the issue is located within the `pkey_rsa_ctrl` function:
```
*(unsigned char **)p2 = rctx->oaep_label;
```
Both `p2` and `rctx->oaep_label` seem to be NULL. It appears that a simple NULL check is missing.

This problem surfaced after the introduction of the following [commit](https://github.com/openssl/openssl/pull/20780/files#diff-910515c87ac6c81bba3f87a197f149cdcc050330964f3ea26877217df4315e0eR2805).
```
ctx.ctrl_cmd = translation->ctrl_num;
```

### PR DESCRIPTION
**Pull request by ljuzwiuk**
Fixes #21335

The `EVP_PKEY_CTX_set0_rsa_oaep_label` and `EVP_PKEY_CTX_get0_rsa_oaep_label` functions have been fixed for usage with the legacy engine. OSSL params are now correctly passed and set in the context.

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
@ljuzwiuk Would you be please able to sign a CLA? https://www.openssl.org/policies/cla.html

**Comment by ljuzwiuk** [comment]
> @ljuzwiuk Would you be please able to sign a CLA? https://www.openssl.org/policies/cla.html

Already sent my ICLA. I am waiting for CCLA to be approved.

File: test/build.info [hunk 1]
```diff
@@ -41,8 +41,6 @@ IF[{- !$disabled{tests} -}]
           evp_pkey_provided_test evp_test evp_extra_test evp_extra_test2 \
           evp_fetch_prov_test evp_libctx_test ossl_store_test \
           v3nametest v3ext punycode_test \
-          evp_pkey_provided_test evp_test evp_extra_test evp_extra_test2 \
-          evp_fetch_prov_test v3nametest v3ext \
```
Comment (t8m) [inline test/build.info:45]: This change should be in a separate commit (can still be in this PR) because it is unrelated.

File: crypto/evp/ctrl_params_translate.c [hunk 2]
```diff
@@ -693,6 +693,9 @@ static int default_fixup_args(enum state state,
                                    translation->param_data_type);
                     return 0;
                 }
+            } else if (state == PRE_PARAMS_TO_CTRL && ctx->action_type == GET) {
+                /* Assign output */
+                ctx->p2 = ctx->params->data;
```
Comment (t8m) [inline crypto/evp/ctrl_params_translate.c:698]: Instead of this change would it work to instead apply :

```diff
--- a/crypto/evp/ctrl_params_translate.c
+++ b/crypto/evp/ctrl_params_translate.c
@@ -2310,7 +2310,7 @@ static const struct translation_st evp_pkey_ctx_translations[] = {
       OSSL_ASYM_CIPHER_PARAM_OAEP_LABEL, OSSL_PARAM_OCTET_STRING, NULL },
     { GET, EVP_PKEY_RSA, 0, EVP_PKEY_OP_TYPE_CRYPT,
       EVP_PKEY_CTRL_GET_RSA_OAEP_LABEL, NULL, NULL,
-      OSSL_ASYM_CIPHER_PARAM_OAEP_LABEL, OSSL_PARAM_OCTET_STRING, NULL },
+      OSSL_ASYM_CIPHER_PARAM_OAEP_LABEL, OSSL_PARAM_OCTET_PTR, NULL },
 
     { SET, EVP_PKEY_RSA, 0, EVP_PKEY_OP_TYPE_CRYPT,
       EVP_PKEY_CTRL_RSA_IMPLICIT_REJECTION, NULL,
```

Comment (ljuzwiuk) [reply crypto/evp/ctrl_params_translate.c:698]: It does not work. However, it does resolve an error that I had overlooked previously. Good catch.

The overlooked error message: 401745F7FF7F0000:error:07800081:common libcrypto routines:set_string_internal:param of incompatible type:crypto/params.c:1371

Comment (t8m) [inline crypto/evp/ctrl_params_translate.c:698]: This adds a few necessary checks, please try it:

```suggestion
            } else if (state == PRE_PARAMS_TO_CTRL && ctx->action_type == GET
                       && translation->param_data_type == OSSL_PARAM_OCTET_PTR) {
                /* Assign output */
                if (!OSSL_PARAM_get_octet_ptr(ctx->params, &ctx->p2, &ctx->sz))
                    return 0;              
```

**Review by t8m — CHANGES_REQUESTED** [review]

File: crypto/evp/ctrl_params_translate.c (hunk 2 above)
Comment (ljuzwiuk) [reply crypto/evp/ctrl_params_translate.c]: After calling `OSSL_PARAM_get_octet_ptr`, `ctx->p2` remains NULL.
To mitigate the potential side effects in other functions, I have refactored the code.

Comment (t8m) [reply crypto/evp/ctrl_params_translate.c]: > After calling `OSSL_PARAM_get_octet_ptr`, `ctx->p2` remains NULL. To mitigate the potential side effects in other functions, I have refactored the code.

This is weird. I thought it should work. Using special fixup function is not right as there are other parameters that should be handled in the exactly same way and currently aren't working.

Comment (t8m) [reply crypto/evp/ctrl_params_translate.c]: Please try this instead:

```diff
diff --git a/crypto/evp/ctrl_params_translate.c b/crypto/evp/ctrl_params_translate.c
index 480d48429b..313f7d2b72 100644
--- a/crypto/evp/ctrl_params_translate.c
+++ b/crypto/evp/ctrl_params_translate.c
@@ -647,6 +647,9 @@ static int default_fixup_args(enum state state,
                                    translation->param_data_type);
                     return 0;
                 }
+            } else if (state == PRE_PARAMS_TO_CTRL && ctx->action_type == GET) {
+                if (translation->param_data_type == OSSL_PARAM_OCTET_PTR)
+                    ctx->p2 = &ctx->bufp;
             } else if ((state == POST_PARAMS_TO_CTRL || state == PKEY)
                        && ctx->action_type == GET) {
                 /* For the POST state, only getting needs some work to be done */
```

Comment (ljuzwiuk) [reply crypto/evp/ctrl_params_translate.c]: That works. Now, I need to find a suitable place to add an extra dereference to obtain the correct value of the label. Specifically, I have to place `ctx->p2 = *(unsigned char**)ctx->p2` within the `POST_PARAMS_TO_CTRL` step. However, it does not seem to fit well.

Comment (t8m) [reply crypto/evp/ctrl_params_translate.c]: > That works. Now, I need to find a suitable place to add an extra dereference to obtain the correct value of the label. Specifically, I have to place `ctx->p2 = *(unsigned char**)ctx->p2` within the `POST_PARAMS_TO_CTRL` step. However, it does not seem to fit well.

You don't. This line handles it https://github.com/openssl/openssl/pull/21401/files#diff-910515c87ac6c81bba3f87a197f149cdcc050330964f3ea26877217df4315e0eL683

Comment (ljuzwiuk) [reply crypto/evp/ctrl_params_translate.c]: > 

We need one more dereference, as without it, I am getting random data in the current state.

Comment (t8m) [reply crypto/evp/ctrl_params_translate.c]: Actually sorry, the line 686 is the right one and the size handling is not correct there either. What about: `return OSSL_PARAM_set_octet_ptr(ctx->params, *((void **)ctx->p2), ret);`

**Review by t8m — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
Merged, thanks for the contribution.
